src/server/utils/bucket_diff.js

```diff
@@ -0,0 +1,464 @@
+    }
+
+    /**
+     * @param {{ keys_diff_map: any; keys_contents_left: any; keep_listing_second_bucket?: boolean; }} ans
+     * @param {string | any[]} etag_pos_on_first_bucket
+     * @param {boolean} version
+     * @param {string} cur_first_bucket_key
+     * @param {any[]} first_bucket_curr_obj
+     * @param {any[]} second_bucket_curr_obj
+     */
+    _etag_not_in_first_bucket(ans, etag_pos_on_first_bucket, version, cur_first_bucket_key, first_bucket_curr_obj, second_bucket_curr_obj) {
+        let pos = 0;
+        let should_continue = false;
+        dbg.log1('ETag is not in the first bucket', etag_pos_on_first_bucket);
+        //     Version: someone wrote directly to the second bucket. need to drill down on the second bucket to see comparison. 
+        //              or the first bucket have a list of max keys and all the keys versions there are newer (with different etag).
+        if (version) {
+            //We will need to drill down on the position of the etag in the second_bucket_curr_obj
+            while (etag_pos_on_first_bucket.length === 0 && pos < second_bucket_curr_obj.length) {
+                pos += 1;
+                etag_pos_on_first_bucket = this._get_etag_pos(pos, second_bucket_curr_obj, first_bucket_curr_obj);
+            }
+            // If non of the ETags of this object exists in the first bucket all the versions are diff.
+            if (etag_pos_on_first_bucket.length === 0 && pos >= second_bucket_curr_obj.length) {
+                // Gap: this is not accurate, if only several versions are newer in the destination, then *some* need to be replicated.
+                //      for now, we will take an assumption that when the latest in the destination is newer we should not consider as diff.
+                this._check_last_modified_for_replication(ans, cur_first_bucket_key, first_bucket_curr_obj, second_bucket_curr_obj);
+                should_continue = true;
+            }
+        } else { // in non version flow - if the second bucket key etag is not the same as the first bucket key etag - this is a diff.
+            // in replication we want to replicate only if the object in the first bucket is more recent then the one in the second bucket
+            this._check_last_modified_for_replication(ans, cur_first_bucket_key, first_bucket_curr_obj, second_bucket_curr_obj);
+            should_continue = true;
+        }
+        return { should_continue, etag_pos_on_first_bucket };
+    }
+
+    _populate_diff_map_and_omit_contents_left(ans, cur_bucket_key, bucket_curr_obj) {
+        ans.keys_diff_map[cur_bucket_key] = bucket_curr_obj;
+        ans.keys_contents_left = _.omit(ans.keys_contents_left, cur_bucket_key);
+    }
+
+    _check_last_modified_for_replication(ans, cur_first_bucket_key, first_bucket_curr_obj, second_bucket_curr_obj) {
+        if (this.for_replication && first_bucket_curr_obj[0].LastModified < second_bucket_curr_obj[0].LastModified) {
```

Comment:
add a comment about why we will omit the key on replication for the future configuration